### PR TITLE
Verify master stack in CI

### PR DESCRIPTION
**Pull request by austinbyers**
## Background

Panther has two different deployment methods - developers do not use the same template that we publish for our users. It's easy to forget parameters in the master template, so now testing will automatically catch these issues immediately instead of waiting until we start the release process. Examples:

> deployments/master.yml: BootstrapGateway: parameter Extra does not exist
deployments/master.yml: Core: missing required parameter Debug
deployments/master.yml: Onboard.Properties.Parameters.AlarmTopicArn: output AlarmTopicARN does not exist in Bootstrap

Closes: #1238 (see issue for more context)

## Changes

- `mage test:cfn` has new checks for the master template (shown above):
    - parameter list must exactly match the parameters defined in the nested stack
    - validate all output references actually exist
- Refactor the very large `test_namespace.go` into separate files: `test_go.go`, `test_python.go`, `test_web.go`, etc. Each file represents its own mage command
- Add space in `mage doc` generated headers (enterprise backport)
- Capture test stdout *and* stderr when applicable - some programs (e.g. `npm run test`) output failures to stdout which we were previously ignoring (enterprise backport)
- Remove `go mod tidy` from `mage fmt` - some libraries may be added even if they are not directly imported by source code (e.g .`xgo` cross-compilation) - we'll let developers decide the appropriate module file instead of automatically flattening it in CI
- Fixes a bug introduced by #1337 - we need to add license headers for *both* the "tools" and "cmd" directories

## Testing

- `mage test:cfn` with different failure scenarios

### CODE REVIEW

**Comment by austinbyers** [comment]
All of the following test files are just code moved from `test_namespace.go` and have not meaningfully changed. No review necessary (unless you want to):

- `test_go.go` (my personal favorite filename)
- `test_web.go`
- `test_integration.go`
- `test_python.go`
- `test_ci.go`

File: tools/mage/test_cfn.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+	errs = append(errs, cfnValidateMaster(parsed)...)
+
+	if len(errs) > 0 {
+		return errors.New(strings.Join(errs, "\n"))
+	}
+	return nil
+}
+
+// Enforce custom resources in a CloudFormation template. Returns a list of error messages.
+func cfnValidateCustomResources(template string, resources map[string]cfnResource) []string {
```
Comment (austinbyers) [inline tools/mage/test_cfn.go:131]: This and `cfnTestFunction` are the same as before, just moved

**Comment by 3nvi** [comment]
Great work @austinbyers . That's super helpful. I've had the same thing created for the FE ENV vars and it has saved me a couple of times 😄

File: tools/mage/master_namespace.go [hunk 2]
```diff
@@ -157,18 +157,26 @@ func masterPackage(bucket, pantherVersion, imgRegistry string) string {
...
+
+	var cfn template
+	if err := cfnparse.ParseTemplate(pythonVirtualEnvPath, "deployments/master.yml", &cfn); err != nil {
 		logger.Fatal(err)
 	}
-	type m = map[string]interface{}
-	version := cfn["Mappings"].(m)["Constants"].(m)["Panther"].(m)["Version"].(string)
 
-	if version == "" {
+	if cfn.Mappings.Constants.Panther.Version == "" {
```
Comment (rleighton) [inline tools/mage/master_namespace.go:175]: Very nice. Much tidier interface,

File: tools/mage/test_cfn.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+		for paramName := range passedParams {
+			if _, exists := parsed[templatePath].Parameters[paramName]; !exists {
+				errs = append(errs, fmt.Sprintf(
+					"deployments/master.yml: %s: parameter %s does not exist",
+					resourceID, paramName))
+			}
+		}
+
+		// Ensure all stack output references are valid. For example,
+		//    !GetAtt Bootstrap.Outputs.AlarmTopicArn
```
Comment (rleighton) [inline tools/mage/test_cfn.go:257]: Do you need to consider `!Ref` ?

File: tools/mage/master_namespace.go (hunk 2 above)
Comment (austinbyers) [reply tools/mage/master_namespace.go:175]: It also won't panic now if one of the intermediate fields is missing

File: tools/mage/test_cfn.go (hunk 3 above)
Comment (austinbyers) [reply tools/mage/test_cfn.go:257]: Nope, `!GetAtt STACK.Outputs.OUTPUT_NAME` is the only way to get stack outputs

**Pull request merged**